sonar-home/src/main/java/org/sonar/home/cache/PersistentCache.java

```diff
@@ -0,0 +1,266 @@
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.concurrent.Callable;
+
+import static java.nio.file.StandardOpenOption.*;
+
+public class PersistentCache {
+  private static final Charset ENCODING = StandardCharsets.UTF_8;
+  private static final String DIGEST_ALGO = "MD5";
+  private static final String LOCK_FNAME = ".lock";
+
+  private Path baseDir;
+
+  // eviction strategy is to expire entries after modification once a time duration has elapsed
+  private final long defaultDurationToExpireMs;
+  private Log log;
+  private boolean forceUpdate;
+
+  public PersistentCache(Path baseDir, long defaultDurationToExpireMs, Log log, boolean forceUpdate) {
+    this.baseDir = baseDir;
+    this.defaultDurationToExpireMs = defaultDurationToExpireMs;
+    this.log = log;
+    this.forceUpdate = forceUpdate;
+
+    log.info("cache: " + baseDir + " default expiration time (ms): " + defaultDurationToExpireMs);
+
+    if (forceUpdate) {
+      log.debug("cache: forcing update");
+    }
+
+    try {
+      Files.createDirectories(baseDir);
+    } catch (IOException e) {
+      throw new IllegalStateException("failed to create cache dir", e);
+    }
+  }
+
+  public Path getBaseDirectory() {
+    return baseDir;
+  }
+
+  public boolean isForceUpdate() {
+    return forceUpdate;
+  }
+
+  @CheckForNull
+  public synchronized String getString(@Nonnull String obj, @Nullable final Callable<String> valueLoader) throws Exception {
+    byte[] cached = get(obj, new Callable<byte[]>() {
+      @Override
+      public byte[] call() throws Exception {
+        String s = valueLoader.call();
+        if (s != null) {
+          return s.getBytes(ENCODING);
+        }
+        return null;
+      }
+    });
+
+    if (cached == null) {
+      return null;
+    }
+
+    return new String(cached, ENCODING);
+  }
+
+  @CheckForNull
+  public synchronized byte[] get(@Nonnull String obj, @Nullable Callable<byte[]> valueLoader) throws Exception {
+    String key = getKey(obj);
+    log.debug("cache: " + obj + " -> " + key);
+
+    try (FileLock l = lock()) {
+      if (!forceUpdate) {
+        byte[] cached = getCache(key);
+
+        if (cached != null) {
+          log.debug("cache hit for " + obj);
+          return cached;
+        }
+
+        log.debug("cache miss for " + obj);
+      }
+      else {
+        log.debug("cache force update for " + obj);
+      }
+
+      if (valueLoader != null) {
+        byte[] value = valueLoader.call();
+        if (value != null) {
+          putCache(key, value);
+        }
+        return value;
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Deletes all cache entries
+   */
+  public synchronized void clear() {
+    log.info("cache: clearing");
+    try (FileLock l = lock()) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused "l" local variable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1481)